### PR TITLE
add python/numpy based solution

### PR DESCRIPTION
**Pull request by emillynge**
## Description
This is a *unfaithful* python implementation that uses numpy array for bit sieve. It is *only* unfaithful due to the numpy dependency.

I still think it is a valuable solution to include due to how widespread numpy is and how close the solution is to the original.

## Contributing requirements

* [x] I read the contribution guidelines in CONTRIBUTING.md.
* [x] I placed my solution in the correct solution folder.
* [x] I added a README.md with the right badge(s).
* [x] I added a Dockerfile that builds and runs my solution.
* [x] I selected `drag-race` as the target branch.
* [x] All code herein is licensed compatible with BSD-3.

### CODE REVIEW

**Comment by tjol** [comment]
How does the performance compare to solution_2? Do you think it might make sense to add this as a variant of solution_2 rather than as a separate solution? They do both use array-based processing, the difference being that solution_2 uses stdlib features with their limitations

File: PrimePython/solution_3/PrimePY.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+
+        # sqrt doesn't seem to make any difference in CPython,
+        # but works much faster than "x**.5" in Pypy
+        q = sqrt(self._size) / 2
+
+
+        bits_view = self._bits[factor:]
+        while factor <= q:
+            for v in bits_view.flat:
```
Comment (tjol) [inline PrimePython/solution_3/PrimePY.py:60]: `np.argwhere`?

**Comment by rbergen** [comment]
> Do you think it might make sense to add this as a variant of solution_2 rather than as a separate solution? They do both use array-based processing, the difference being that solution_2 uses stdlib features with their limitations

Due to the fact that this implementation constitutes a different approach, and solution_2 was created by a different author, I am okay with adding this is as a separate solution.

File: PrimePython/solution_3/PrimePY.py (hunk 1 above)
Comment (emillynge) [reply PrimePython/solution_3/PrimePY.py:60]: Unfortunately, `argwhere` will return **all** indices to `True` entries - which will be close to `N` in the first iterations trending towards zero.

I looked for a long time, but numpy has *no* equivalent to `list.index`, which will return just the first occurence. I find that very odd, but there really uis no good solution.

`argmax` is the closest fit, but performs much worse than the current loop.

Given that the maximum number of iterations in the while loop is around 14-17, the problem performance hit is not too bad.

**Comment by emillynge** [comment]
If the original author is OK with it, then I would be perfectly happy to merge my implementation into solution_2. They do after all share around 90-95% of the code.

I could just add `--with-numpy` flag that uses numpy for the bits array, and let the docker container execute the program twice with different flags.

There is an issue though, with the github tags since my implementationm is not faithful, whereas solution_2 is.

File: PrimePython/solution_3/PrimePY.py (hunk 1 above)
Comment (tjol) [reply PrimePython/solution_3/PrimePY.py:60]: Right! Makes sense.

**Comment by tjol** [comment]
@rbergen Just to be clear, I didn't mean to step on your toes, I was just thinking aloud.

@emillynge What I meant to suggest was, if anything, to consider adding your program as a separate file to the solution_2 directory. But really it's unnecessary, and may just confuse things.

**Comment by emillynge** [comment]
Regarding performance, as you can see from the included benchmark in the readme; Using numpy increases performance by 360% over solution_2 which leaves it at roughly 60% of the performance of PrimeC/solution_1!!

**Comment by tjol** [comment]
Good lord, was that table right in front of me this whole time? I should really read things more carefully before asking stupid questions next time :-)

Very cool!

**Comment by rbergen** [comment]
> @rbergen Just to be clear, I didn't mean to step on your toes, I was just thinking aloud.

@tjol I have decently-sized feet, but my toes aren't **that** long. :)

**Review by rbergen — APPROVED** [review]
LGTM

**Comment by emillynge** [comment]
BTW, full disclosure, just to be clear.

The numpy dependency relies very heavily on C extensions and the speedup is most certainly due to critical parts being executed in C extensions.

The difference between this PR and #370 is mostly that the programmer doesn't need to know any C at all to write the program.
In fact, most users of numpy likely does not know that it is very much a C library.

File: PrimePython/solution_3/PrimePY.py (hunk 1 above)
Comment (mgdotdev) [reply PrimePython/solution_3/PrimePY.py:60]: If you want to, you could do this in C by iterating across the numpy array until you find the first `1`:
```C
#include <Python.h>
#include <numpy/arrayobject.h>

static PyObject* _find_first(PyObject *self, PyObject *args) {
    PyArrayObject* array;
    int item;

    if (!PyArg_ParseTuple(args, "Oi", &array, &item)) {
        return NULL;
    }

    npy_intp _len = PyArray_DIM(array, 0);
    npy_intp _stride = PyArray_STRIDE(array, 0);

    for (int i=0; i<_len; i++) {
        if (*(int*)(array->data+i*_stride) == item) {
            return Py_BuildValue("i", i);
        }
    }

    return Py_BuildValue("i", -1);
}

static char indexMethodsDocs[] = 
    "return the index of first instance of a provided integer value within a numpy array";


static PyMethodDef indexMethods[] = {
    {"find_first", _find_first, METH_VARARGS, indexMethodsDocs},
    {NULL, NULL, 0, NULL}
};

static struct PyModuleDef indexModule = {
    PyModuleDef_HEAD_INIT,
    "indexModule",
    NULL,
    -1,
    indexMethods
};

PyMODINIT_FUNC PyInit_indexModule(void) {
    return PyModule_Create(&indexModule);
}
```

which you could then call from python (after you set it up as an extension) via

```python
index = indexModule.find_first(array, 1)
```

Comment (emillynge) [reply PrimePython/solution_3/PrimePY.py:60]: Even though I am calling numpy which I know uses C extensions i'd rather not write any actual C code for this solution.

What I'm trying to show, is that by clever use of a widely available package, you can get C-like performance without writing a single line of C yourself.
Also, keeping the program compile-free is nice.

**Review by marghidanu — APPROVED** [review]

**Pull request merged**